xla/service/gpu/fusions/transpose_mlir.cc

```diff
@@ -405,5 +410,376 @@ IndexingMap MlirTransposeFusion::GetIndexing(bool input,
+        bool reuse_operand =
+            ref_desc.ref_count == 0 && output_bytes <= ref_desc.elem_bytes;
+        can_reuse = ref_desc.tile_size > max_tile_size
+                        ? reuse_operand
+                        : can_reuse || reuse_operand;
+        if (reuse_operand) {
+          reuse_desc_id = operand_desc.ref_id;
+        }
+        max_tile_size = ref_desc.tile_size;
+      }
+    }
+    if (can_reuse) {
+      auto& reuse_desc = (*instr_bytes_desc)[reuse_desc_id];
+      ++reuse_desc.ref_count;
+      unused_bytes -= reuse_desc.tile_size * reuse_desc.elem_bytes;
+      return HloBytesUsageDesc{0, reuse_desc.elem_bytes, 0, reuse_desc_id,
+                               unused_bytes};
+    }
+    return HloBytesUsageDesc{max_tile_size, output_bytes, 1,
+                             instr_to_ids[instr], unused_bytes};
+  }
+  // We can add more supported hlo if needed.
+  return std::nullopt;
+}
+
+std::optional<int> ComputeMaxBytesCountFromInstrs(
+    absl::Span<const HloInstructionAdaptor> post_order,
+    absl::flat_hash_map<const HloInstruction*, int>& instr_to_ids,
+    absl::InlinedVector<int, 2>& live_range,
+    absl::InlinedVector<HloBytesUsageDesc, 2>* instr_bytes_desc,
+    absl::InlinedVector<bool, 2>* visited,
+    const std::function<bool(const HloInstruction*)>& should_visit,
+    int instrs_bytes) {
+  int max_bytes_count = instrs_bytes;
+  for (auto instr_adaptor : post_order) {
+    auto* instr = &instr_adaptor.instruction();
+    int instr_id = instr_to_ids[instr];
+    if (!should_visit(instr) || (*visited)[instr_id]) {
+      continue;
+    }
+    bool operands_visited = true;
+    for (auto operand_adaptor : instr_adaptor.GetOperands()) {
+      auto* operand = &operand_adaptor.instruction();
+      int operand_id = instr_to_ids[operand];
+      operands_visited = operands_visited && (*visited)[operand_id];
+    }
+    if (operands_visited) {
+      int instr_id = instr_to_ids[instr];
+      auto bytes_desc = GetHloBytesUsageDesc(instr_adaptor, live_range,
+                                             instr_bytes_desc, instr_to_ids);
+      if (!bytes_desc.has_value()) {
+        return std::nullopt;
+      }
+      (*instr_bytes_desc)[instr_id] = bytes_desc.value();
+      (*visited)[instr_id] = true;
+      instrs_bytes +=
+          bytes_desc.value().tile_size * bytes_desc.value().elem_bytes;
+      max_bytes_count = std::max(max_bytes_count, instrs_bytes);
+      instrs_bytes -= bytes_desc.value().unused_bytes;
+      // This means current hlo is the root of fusion. We release the occupied
+      // registers immediately.
+      if (live_range[instr_id] == instr_id) {
+        auto& ref_desc = (*instr_bytes_desc)[bytes_desc.value().ref_id];
+        --ref_desc.ref_count;
+        if (ref_desc.ref_count == 0) {
+          instrs_bytes -= ref_desc.tile_size * ref_desc.elem_bytes;
+        }
+      }
+    }
+  }
+  return max_bytes_count;
+}
+
+std::optional<int> ComputeRegNumFromParamToTarget(
```

Comment:
I don't really understand why indexing maps should be related to register pressure at all. Say we have a reduction in the prologue: the register use *might* depend a bit on the size of the reduction, but not significantly. For large reductions, it doesn't depend on it at all. For small reductions, where the loop is unrolled, we get something like

```
reducer(reducer(reducer(reducer(init, e[0]), e[1]), e[2]), e[3])
```

As you can see, this shouldn't need signficantly more registers than reducer(init, e[0]) on its own, since no intermediates need to be kept around.

Reply:
I focus on the indexing map of parameters but not other. For the above case you provide, I don't compute the registers usage of intermediates. And the indexing map of parameter help us to distinguish these two cases:
```
parameter -> transpose
parameter -> broadcast -> transpose
```
The second case need less registers than the first. And the indexing map is special, we only compute loaded elements in one block, you can see this line https://github.com/openxla/xla/pull/14759/files#diff-0210cd152294b1475469517ffba6b5ee578acc4e786b0dceeee6854aba445a29R647.